man/fi_peer.3.md

```diff
@@ -286,6 +286,91 @@ on the local node.
+An owner may setup peer relationships with multiple providers.
+
+Peer counters are configured by the owner calling the peer’s fi_cntr_open() call.
+ The owner passes in the FI_PEER flag to fi_cntr_open().
+When FI_PEER is specified, the context parameter passed into
+fi_cntr_open() must reference a struct fi_peer_cntr_context.
+Providers that do not support peer counters must fail the
+fi_cntr_open() call with -FI_EINVAL
+(indicating an invalid flag). The fid_peer_cntr referenced
+by struct fi_peer_cntr_context must remain valid until the peer’s counter is closed.
+
+The data structures to support peer counters are defined as follows:
+
+```c
+struct fi_ops_cntr_owner {
+    size_t  size;
+    ssize_t (*inc)(struct fid_peer_cntr *cntr);
+    ssize_t (*incerr)(struct fid_peer_cntr *cntr);
+};
+
+struct fid_peer_cntr {
+    struct fid fid;
+    struct fi_ops_cntr_owner *owner_ops;
+};
+
+struct fi_peer_cntr_context {
+    size_t size;
+    struct fid_peer_cntr *cntr;
+};
+```
+
+For struct fid_peer_cntr, the owner initializes the fid and owner_ops fields.
+struct fi_ops_cntr_owner is used by the peer to communicate with the owning provider.
+
+If manual progress is needed on the peer counter,
+the owner should drive progress by using the fi_cntr_read()
+and the fi_cntr_read() should do nothing but progress the peer cntr.
+The peer provider should set other functions that attempt to access
+the peer’s cntr (i.e. fi_cntr_readerr, fi_cntr_set, etc.) to return -FI_ENOSYS.
+
+## fi_ops_cntr_owner::inc()
+
+This call directs the owner to increment the value of the cntr.
+
+## fi_ops_cntr_owner::incerr()
+
+The behavior of this call is similar to the inc() ops. It increments the error
+of the cntr indicating that a data transfer has failed into the cntr.
+
+
+## EXAMPLE COUNTER SETUP
+
+The above description defines the generic mechanism for sharing cntrs between providers. This section outlines one possible implementation to demonstrate the use of the APIs. In the example, provider A uses provider B as a peer for data transfers targeting endpoints on the local node.
+
+```
+1. Provider A is configured to use provider B as a peer.  This may be coded
+   into provider A or set through an environment variable.
+2. The application calls:
+   fi_cntr_open(domain_a, attr, &cntr_a, app_context)
+3. Provider A allocates cntr_a and automatically configures it to be used
+   as a peer cntr.
+4. Provider A takes these steps:
+   allocate peer_cntr and reference cntr_a
+   set peer_cntr_context->cntr = peer_cntr
+   set attr_b.flags |= FI_PEER
+   fi_cntr_open(domain_b, attr_b, &cntr_b, peer_cntr_context)
+5. Provider B allocates a cntr, but configures it such that all completions
+   are added to the peer_cntr.  The cntr ops to read from the cntr are
+   set to enosys calls.
+6. Provider B inserts its own callbacks into the peer_cntr object.  It
+   creates a reference between the peer_cntr object and its own cntr.
+```
```

Comment:
Most of the above text is a copy-paste-modify of the peer CQ.  We don't need to duplicate all the text.  See the peer domain, eq, etc. below for how to simplify the text and simply reference the peer CQ description.

Reply:
Simplified in the new revision